### PR TITLE
bpo-30803: clarify truth value testing documentation

### PR DESCRIPTION
**Pull request by peterthomassen**
The documentation was unclear about the truth value of sets. This PR clarifies the situation, based on a phrasing proposal from https://bugs.python.org/issue30803.

https://bugs.python.org/issue30803

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

File: Doc/library/stdtypes.rst
Comment (terryjreedy) [inline Doc/library/stdtypes.rst]: * any empty set or mapping, for example, ``set()``, ``{}``.

File: Doc/library/stdtypes.rst
Comment (terryjreedy) [inline Doc/library/stdtypes.rst]: I much prefer a minimal change, adding "set or" and "set(), "

**Review by terryjreedy — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
This PR has both the "CLA signed" and the "CLA not signed" labels. Is it Schrödinger's CLA? ;)

**Comment by peterthomassen** [comment]
@terryjreedy Well, I'm not sure what to do at this point. I adapted the [proposal](https://bugs.python.org/issue30803#msg297312) by @rhettinger who is also a very experienced Python contributor. I found this a very reasonable course of action.

Now you say that you "much prefer a minimal change" while putting forward zero arguments (not even a subjective one). On bpo, you also said very broadly that you would not merge this patch. It's true that patches don't need to be accepted, but it's also true that a community works better if a contribution is valued, and criticism is backed by an argument.

Other people may "much prefer" concise documentation. The previous phrasing was not concise, and I can't following the reasoning for keeping something just because the change would be minimal. In fact, one may argue that this is a good moment to rephrase the whole thing, as we're touching it anyways.

File: Doc/library/stdtypes.rst
Comment (terryjreedy) [inline Doc/library/stdtypes.rst]: I strongly prefer 'collection' rather than 'container'.  Collections do not 'contain' python objects.  They may contain references to objects, values corresponding to objects, or info needed to generate objects by algorithm.

Comment (terryjreedy) [reply Doc/library/stdtypes.rst]: That aside, Raymond's wording is clearer, simpler,  and more complete than your rewrite.  Use what he wrote.

Comment (peterthomassen) [reply Doc/library/stdtypes.rst]: I don't see why one can't say 'container' even if the structure may 'contain' entities beyond Python objects. There is still something contained within; further, nobody claimed them to be 'object containers' only or made some other implicit restriction.

Similarly: Does a 'collection' only 'collect' Python objects? (Does it actually 'collect' anything?)

Anyhow, the docs use 'collection' and 'container' pretty much interchangeably, and I consider this point completely moot. However, you certainly have better understanding, so I used now what you strongly and wisely preferred, hence graciously commanded me to. I hope we now all agree on this matter.

Comment (bitdancer) [reply Doc/library/stdtypes.rst]: @peterthomassen just FYI, 'collection' is technically correct here and 'container' is technically wrong, since the Container abc only requires `__contains__` to be defined, which if nothing else was defined would make the object true.  A Collection, on the other hand, is sized, and so can be false if empty.

You are right, however, that we do sometimes use them interchangeably.  But we probably shouldn't :)

Thanks for taking the time to work through this PR.

Comment (peterthomassen) [reply Doc/library/stdtypes.rst]: @bitdancer Thank you for sharing your knowledge. I understand your point (which is technical, not just semantic), and agree that as a consequence, 'empty container' is not a very good term if a `Container` type is not required to have `__len__` defined. (As an aside, `isinstance(set(), collections.abc.Container)` is still true, and the same is the case for the other empty collections listed in the docs section this PR deals with.)

I guess it is not defined whether 'container' and 'collection' mean `Container` and `Collection`, respectively (but I don't feel this needs to be rectified).

**Review by terryjreedy — APPROVED** [review]
I like this version ;-).

File: Doc/library/stdtypes.rst
Comment (bitdancer) [reply Doc/library/stdtypes.rst]: Right, the lower case terms do not necessarily refer to the ABC definitions, but the closer we adhere to those definitions in the docs, the better, probably.  The ABCs represent our formal definitions of the terms, as opposed to the colloquial usage.

Comment (terryjreedy) [reply Doc/library/stdtypes.rst]: I am fussy about the terminology for the practical reason that I regularly see people confused by 'container', including myself as a student long ago when I realized that a set is *not* like a box or room but a pure conceptual grouping.  The issue comes up at least yearly on python-list, where I have been active for over 20 years.

File: Doc/library/stdtypes.rst
Comment (vadmium) [inline Doc/library/stdtypes.rst]: What was wrong with the original wording? And aren’t Decimal and Fraction out of scope of “the built-in objects”?

Comment (terryjreedy) [reply Doc/library/stdtypes.rst]: See the issue for 'wrong'.  Decimal and Fraction are built-in in the sense of C-coded stdlib objects, but not in 'part of builtins' module.  I could go either way with them.

**Review by rhettinger — APPROVED** [review]

File: Doc/library/stdtypes.rst
Comment (vadmium) [reply Doc/library/stdtypes.rst]: Okay, I see that this wording probably comes from @terryjreedy’s suggestion <https://bugs.python.org/issue30803#msg298099>, although I don’t see any reasoning. IMO the current _zero of any numeric type_, or Raymond’s _numbers equal to zero_, are clearer.

_Numeric zero of any type_ raises the question of what a numeric zero of the string type looks like. I think other languages consider strings with the digit zero to be false. (PHP or Perl perhaps?)

Comment (terryjreedy) [reply Doc/library/stdtypes.rst]: I agree with changing back to 'zero of any numeric type'.  Peter, change this (I cannot, for some reason) and I will merge.

**Comment by terryjreedy** [comment]
Peter, I cannot add commits to this issue, and a few others. Please follow the instructions at https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/#enabling-repository-maintainer-permissions-on-existing-pull-requests to make sure "Allow edits from maintainers." is checked for this and other IDLE issues. Please let me know what you find. If this permission is already granted, then I have some other problem with github.

**Comment by peterthomassen** [comment]
@terryjreedy Terry, I just checked that "allow edits" box.

**Pull request merged**
